### PR TITLE
chore(core): implement allocation-free base64 decoder

### PR DESCRIPTION
**Pull request by jerrinot**
It's useful in the secret store

also added a code for the base64 URL variant (a slightly different alphabet and no padding)
all base64 functionality moved from `Chars` to `Encoding`

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 78 / 80 (97.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/Encoding.java|64|66|96.97%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/UserDatabasePgWireAuthenticationFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Chars.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/auth/AuthUtils.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64FunctionFactory.java|3|3|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 77 / 80 (96.25%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/Encoding.java|63|66|95.45%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/UserDatabasePgWireAuthenticationFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Chars.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/auth/AuthUtils.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64FunctionFactory.java|3|3|100.00%|

File: core/src/main/java/io/questdb/std/Encoding.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            char letter = alphabet[i];
+            assert letter < 128;
+            inverted[letter] = (byte) i;
+        }
+        return inverted;
+    }
+
+    private static byte invertedLookup(byte[] invertedAlphabet, char ch) {
+        if (ch > 127) {
+            throw CairoException.nonCritical().put("not ascii letter while decoding base64 [ch=").put(Character.getNumericValue(ch)).put(']');
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/std/Encoding.java:139]: nit:
```suggestion
            throw CairoException.nonCritical().put("non-ascii character while decoding base64 [ch=").put(Character.getNumericValue(ch)).put(']');
```

**Review by puzpuzpuz — DISMISSED** [review]

**Comment by jerrinot** [comment]
@puzpuzpuz thanks for the review, I pushed some minor changes. (I have an idea how to improve performance of the decoder, but that will require more work with benchmark so I'm postponing it.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 78 / 78 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/UserDatabasePgWireAuthenticationFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Chars.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/std/Encoding.java|64|64|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/auth/AuthUtils.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64FunctionFactory.java|3|3|100.00%|

**Review by puzpuzpuz — DISMISSED** [review]

File: core/src/main/java/io/questdb/std/Encoding.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
+package io.questdb.std;
+
+import io.questdb.cairo.CairoException;
+import io.questdb.std.str.CharSink;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+
+public final class Encoding {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/std/Encoding.java:9]: would it be possible to keep this in `Chars.` There is refactoring moving `encoding` to another class, which makes it hard to see changes. Also `Encoding` is too generic, basically as generic as `Chars`.

Generally there is nothing wrong with having `Encoding`, but I am trying to address utility class prolifiration, which makes it harder to locate utility methods

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 78 / 78 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/UserDatabasePgWireAuthenticationFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Chars.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/std/Encoding.java|64|64|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/auth/AuthUtils.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64FunctionFactory.java|3|3|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 92 / 92 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/UserDatabasePgWireAuthenticationFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Chars.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/std/Encoding.java|78|78|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/auth/AuthUtils.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64FunctionFactory.java|3|3|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 85 / 85 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/UserDatabasePgWireAuthenticationFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Chars.java|79|79|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/auth/AuthUtils.java|4|4|100.00%|

File: core/src/main/java/io/questdb/std/Encoding.java (hunk 2 above)
Comment (jerrinot) [reply core/src/main/java/io/questdb/std/Encoding.java]: done

**Comment by jerrinot** [comment]
@puzpuzpuz I pushed an optimized version of the decoder. it separates the hot loop where the decoder can always read 4 chars from the last 1-3 chars which requires more complexity. review appreciated!

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 85 / 85 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/UserDatabasePgWireAuthenticationFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Chars.java|79|79|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/auth/AuthUtils.java|4|4|100.00%|

**Comment by puzpuzpuz** [comment]
Benchmark results on OpenJDK 17.0.7, x86-64:
```
Benchmark                                                     Mode  Cnt       Score      Error   Units
Base64Benchmark.jdkBaseline                                   avgt    3   10954.389 ±  653.352   ns/op
Base64Benchmark.jdkBaseline:·gc.alloc.rate                    avgt    3   12680.932 ±  760.669  MB/sec
Base64Benchmark.jdkBaseline:·gc.alloc.rate.norm               avgt    3  152959.293 ±    1.317    B/op
Base64Benchmark.jdkBaseline:·gc.churn.G1_Eden_Space           avgt    3   12775.953 ±  928.301  MB/sec
Base64Benchmark.jdkBaseline:·gc.churn.G1_Eden_Space.norm      avgt    3  154105.268 ± 3317.555    B/op
Base64Benchmark.jdkBaseline:·gc.churn.G1_Survivor_Space       avgt    3       0.006 ±    0.015  MB/sec
Base64Benchmark.jdkBaseline:·gc.churn.G1_Survivor_Space.norm  avgt    3       0.071 ±    0.184    B/op
Base64Benchmark.jdkBaseline:·gc.count                         avgt    3     875.000             counts
Base64Benchmark.jdkBaseline:·gc.time                          avgt    3     322.000                 ms
Base64Benchmark.questDB                                       avgt    3   58496.700 ± 1278.410   ns/op
Base64Benchmark.questDB:·gc.alloc.rate                        avgt    3      ≈ 10⁻⁴             MB/sec
Base64Benchmark.questDB:·gc.alloc.rate.norm                   avgt    3       0.003 ±    0.001    B/op
Base64Benchmark.questDB:·gc.count                             avgt    3         ≈ 0             counts
```

TL;DR we're zero allocation, but 5x slower than the standard library.

**Comment by jerrinot** [comment]
@puzpuzpuz that makes sense as base64 impl in JDK17 running on Intel uses intrinsic. Given I don't expect us to decode huge blocks of base64 encoded blobs I'd prioritize no allocation over higher throughput. wdyt?

**Review by puzpuzpuz — APPROVED** [review]

**Comment by puzpuzpuz** [comment]
> I'd prioritize no allocation over higher throughput. wdyt?

Yes, I agree. I don't see any low-hanging fruits to speed up this code, but I think it's fine.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 87 / 87 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/UserDatabasePgWireAuthenticationFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/FactoryProvider.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Chars.java|79|79|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/auth/AuthUtils.java|4|4|100.00%|

**Comment by jerrinot** [comment]
@puzpuzpuz thanks for the review, all comments addressed

**Comment by jerrinot** [comment]
results on my M1 Mac:
```
Benchmark                                                     (size)  Mode  Cnt       Score       Error   Units
Base64Benchmark.jdkBaseline                                       16  avgt    3      37.543 ±     2.006   ns/op
Base64Benchmark.jdkBaseline:·gc.alloc.rate                        16  avgt    3    1740.173 ±    80.739  MB/sec
Base64Benchmark.jdkBaseline:·gc.alloc.rate.norm                   16  avgt    3      72.005 ±     0.001    B/op
Base64Benchmark.jdkBaseline:·gc.churn.G1_Eden_Space               16  avgt    3    1740.294 ±   275.270  MB/sec
Base64Benchmark.jdkBaseline:·gc.churn.G1_Eden_Space.norm          16  avgt    3      72.010 ±     8.133    B/op
Base64Benchmark.jdkBaseline:·gc.churn.G1_Survivor_Space           16  avgt    3       0.004 ±     0.039  MB/sec
Base64Benchmark.jdkBaseline:·gc.churn.G1_Survivor_Space.norm      16  avgt    3      ≈ 10⁻⁴                B/op
Base64Benchmark.jdkBaseline:·gc.count                             16  avgt    3     188.000              counts
Base64Benchmark.jdkBaseline:·gc.time                              16  avgt    3     138.000                  ms
Base64Benchmark.jdkBaseline                                      512  avgt    3     513.483 ±   226.605   ns/op
Base64Benchmark.jdkBaseline:·gc.alloc.rate                       512  avgt    3    2178.626 ±   947.240  MB/sec
Base64Benchmark.jdkBaseline:·gc.alloc.rate.norm                  512  avgt    3    1232.083 ±     0.030    B/op
Base64Benchmark.jdkBaseline:·gc.churn.G1_Eden_Space              512  avgt    3    2178.497 ±  1086.927  MB/sec
Base64Benchmark.jdkBaseline:·gc.churn.G1_Eden_Space.norm         512  avgt    3    1231.950 ±   130.431    B/op
Base64Benchmark.jdkBaseline:·gc.churn.G1_Survivor_Space          512  avgt    3       0.007 ±     0.054  MB/sec
Base64Benchmark.jdkBaseline:·gc.churn.G1_Survivor_Space.norm     512  avgt    3       0.004 ±     0.030    B/op
Base64Benchmark.jdkBaseline:·gc.count                            512  avgt    3     229.000              counts
Base64Benchmark.jdkBaseline:·gc.time                             512  avgt    3     181.000                  ms
Base64Benchmark.jdkBaseline                                    65536  avgt    3   63277.678 ±  3387.728   ns/op
Base64Benchmark.jdkBaseline:·gc.alloc.rate                     65536  avgt    3    2193.151 ±   114.986  MB/sec
Base64Benchmark.jdkBaseline:·gc.alloc.rate.norm                65536  avgt    3  152962.579 ±     2.069    B/op
Base64Benchmark.jdkBaseline:·gc.churn.G1_Eden_Space            65536  avgt    3    2214.139 ±   282.059  MB/sec
Base64Benchmark.jdkBaseline:·gc.churn.G1_Eden_Space.norm       65536  avgt    3  154428.659 ± 25999.283    B/op
Base64Benchmark.jdkBaseline:·gc.churn.G1_Survivor_Space        65536  avgt    3       0.307 ±     0.070  MB/sec
Base64Benchmark.jdkBaseline:·gc.churn.G1_Survivor_Space.norm   65536  avgt    3      21.402 ±     5.212    B/op
Base64Benchmark.jdkBaseline:·gc.count                          65536  avgt    3     236.000              counts
Base64Benchmark.jdkBaseline:·gc.time                           65536  avgt    3     177.000                  ms
Base64Benchmark.questDB                                           16  avgt    3      32.033 ±     1.374   ns/op
Base64Benchmark.questDB:·gc.alloc.rate                            16  avgt    3      ≈ 10⁻⁴              MB/sec
Base64Benchmark.questDB:·gc.alloc.rate.norm                       16  avgt    3      ≈ 10⁻⁶                B/op
Base64Benchmark.questDB:·gc.count                                 16  avgt    3         ≈ 0              counts
Base64Benchmark.questDB                                          512  avgt    3     729.311 ±    17.599   ns/op
Base64Benchmark.questDB:·gc.alloc.rate                           512  avgt    3      ≈ 10⁻⁴              MB/sec
Base64Benchmark.questDB:·gc.alloc.rate.norm                      512  avgt    3      ≈ 10⁻⁴                B/op
Base64Benchmark.questDB:·gc.count                                512  avgt    3         ≈ 0              counts
Base64Benchmark.questDB                                        65536  avgt    3   83206.038 ± 16317.858   ns/op
Base64Benchmark.questDB:·gc.alloc.rate                         65536  avgt    3      ≈ 10⁻⁴              MB/sec
Base64Benchmark.questDB:·gc.alloc.rate.norm                    65536  avgt    3       0.004 ±     0.002    B/op
Base64Benchmark.questDB:·gc.count                              65536  avgt    3         ≈ 0              counts
```

**Review by puzpuzpuz — APPROVED** [review]

**Comment by ideoma** [comment]
@jerrinot does "DO NOT MERGE" tag still apply?

**Comment by jerrinot** [comment]
@ideoma: yes, there is an EE counter-part PR which must be merged at the same time and it's yet to be reviewed.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 87 / 87 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/UserDatabasePgWireAuthenticationFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/FactoryProvider.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Chars.java|79|79|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/auth/AuthUtils.java|4|4|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 87 / 87 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/UserDatabasePgWireAuthenticationFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/FactoryProvider.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Chars.java|79|79|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/auth/AuthUtils.java|4|4|100.00%|
